metals/src/main/scala/scala/meta/internal/metals/debug/SourcePathAdapter.scala

```diff
@@ -17,12 +18,26 @@ private[debug] final class SourcePathAdapter(
     workspace: AbsolutePath,
     sources: Set[AbsolutePath],
     buildTargets: BuildTargets,
-    saveJarFileToDisk: Boolean
+    supportVirtualDocuments: Boolean
 ) {
+  // when virtual documents are supported there is no need to save jars on disk
+  private val saveJarFileToDisk = !supportVirtualDocuments
   private val dependencies = workspace.resolve(Directories.dependencies)
+
   def toDapURI(sourcePath: AbsolutePath): Option[URI] = {
     if (sources.contains(sourcePath)) Some(sourcePath.toURI)
-    else {
+    else if (supportVirtualDocuments) {
+      val jarPath = workspace.toNIO.relativize(sourcePath.toNIO)
+      val x = jarPath.toString
+        .stripPrefix("jar:")
+        .replace("%3A", ":")
+        .replace("%21", "")
+      pprint.log(x)
+      val uri = new URI(x)
+
+      pprint.log(uri)
+      Some(uri)
```

Comment:
I tried here multiple approaches but none of them worked. I'm still trying to figure out how Metals should map virtual doc's URI to the DAP one.
If I'm correct then [here](https://github.com/microsoft/java-debug/blob/625fe84cef4825a4b6b476d04383f5168e81708d/com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java#L284) `ISourceLookUpProvider` is used to convert given uri into a fqcn. But I cannot infer what kind of URI [`SourceLookUpProvider`](https://github.com/scalacenter/scala-debug-adapter/blob/53a062a71c280da15ccaf1cc1d365a03be8bdaa9/core/src/main/scala/ch/epfl/scala/debugadapter/internal/SourceLookUpProvider.scala#L42) expects. Can I ask for some clue @adpi2 ?

Reply:
Previously it was easier because Metals were copying source files from jars to the `.metals/dependecies` directory but now we don't do that anymore (if the editor supports virtual docs) so here we should create a URI which will point to the file inside a JAR.

Reply:
Actually nevermind, I found what I have to do.

Reply:
@kpodsiad I've updated the `URIEncoderDecoder` [here](https://github.com/Arthurm1/metals/tree/uri_decoding)
You should be able to grab that and run `URIEncoderDecoder.decode` on the `jar:file%3A` uri rather than picking out specific chars.
It's VSCode that's encoding the uri this way as it will encode any `:` after the scheme `:`.
Java should be able to open a jar using a uri of the format `jar:file:///somepath/somejar.jar!/somepackage/somefile.java` which (hopefully) is what you will get from that method.  So I'd expect/hope that the debugger will cope with that uri.
Of course the debugger might not understand `jar` uris at all because before virtual docs, all `jar` files were saved in `readonly/jarName/package/class` and I don't know where the code that handles that is.

Reply:
> I've updated the URIEncoderDecoder [here](https://github.com/Arthurm1/metals/tree/uri_decoding)
You should be able to grab that and run URIEncoderDecoder.decode on the jar:file%3A uri rather than picking out specific chars.

Thanks, that might be helpful!

> Of course the debugger might not understand jar uris at all because before virtual docs, all jar files were saved in readonly/jarName/package/class and I don't know where the code that handles that is.

Funny thing, the code is right below my "changes". I switched back to the old Metals server version and it turns out that debugger understands only `jar` uris. But now I can see what should I do so this PR should be relatively easy to finish.

Reply:
> it turns out that debugger understands only jar uris.

Yep, that's correct :slightly_smiling_face: